{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "README\n",
    "\n",
    "**01 Run SNP extraction background.ipynb**: Notebook to run HAIL in the background. Follow the instructions in this notebook.\n",
    "\n",
    "**01 Step 1 of extract PC SNPs from reference 1KGenomes and HGMD**: Create a QC'd HAIL genotype matrix table containing the PC SNPs\n",
    "\n",
    "**01 Step 1 of extract PC SNPs from AoU alpha3**: Create a QC'd HAIL genotype matrix table containing the PC SNPs\n",
    "\n",
    "**01 Step 1 of extract PRS SNPs from AOU alpha3**: Create a QC'd HAIL genotype matrix table containing the PRS SNPs \n",
    "\n",
    "**02 Step 1 of extract PRS SNPs from reference 1KGenomes and HGMD**: Create a QC'd HAIL genotype matrix table containing the PRS SNPs \n",
    "\n",
    "**01 Step 2 extract PC SNPs from AoU alpha3**: Read in QC'd hail genotype matrix from step 1, output separate PLINK files by chr\n",
    "\n",
    "**01 Step 2 extract PC SNPs from reference 1KGenomes and HGMD**: Read in QC'd hail genotype matrix from step 1, output separate PLINK files by chr\n",
    "\n",
    "**01 Step 2 extract PRS SNPs from AoU alpha3**: Read in QC'd hail genotype matrix from step 1, output separate PLINK files by chr\n",
    "\n",
    "**01 Step 2 extract PRS SNPs from reference 1KGenomes and HGMD**: Read in QC'd hail genotype matrix from step 1, output separate PLINK files by chr\n",
    "\n",
    "**02 Calculate PRS alpha3 AoU participants.ipynb**: Calcuate the PRS in _alpha3_ participants, accounting for mismatches, strand flips and duplicate positions. Count number of duplicate positions, mismatches, flips and reverse strands.\n",
    "\n",
    "\n",
    "**02 Calculate PRS Reference.ipynb**: Calcuate the PRS in reference samples, accounting for mismatches, strand flips and duplicate positions. Count number of duplicate positions, mismatches, flips and reverse strands.\n",
    "\n",
    "**Notes on 03 Extract ...**: The Notebooks that extract data were created by using the following steps\n",
    "    \n",
    "    In the Data tab, click on the Dataset plus sign (create a new dataset)\n",
    "    Select the cohort (I selected all WGS participants)\n",
    "    Select the Concept sets of Interest (see additional notes)\n",
    "    Select the columns of interest\n",
    "    Look at the preview of the data \n",
    "    Click on Create Dataset\n",
    "    Click on Analyze (I chose to use R for the code)\n",
    "\n",
    "**Addition Notes on Concept Sets**: Some concept sets already exists such as Demographics and Zip Code Socioeconomic Status. However, I needed to create concept sets for the specific phenotypes that I wanted to pull from the electronic health records. I found that it was best to create each related concept set separately, hence the multiple notebooks. I created the concept sets using the following steps\n",
    "\n",
    "    Begin as above to extract a Dataset\n",
    "    Rather than selecting a concept set, click on the plus sign. This will allow you to create a concept set\n",
    "    Click on the type of phenptype data you need\n",
    "    Use codes or search terms to get a list of potential data types\n",
    "    You can click on the heirarchy to click on individual components if you don't want all component\n",
    "    Click on the green plus sign to add to the concept set\n",
    "    When you have all the concepts, click on Finish & Review\n",
    "    Then click Save Concept Set\n",
    "   \n",
    "You can edit the concept sets after the fact, as well. \n",
    "Currently the concept sets I created are not in the production workspace since I created them in the pre-production workspace. You can refer to the supplemental table of Rosenthal et. al. (https://www.medrxiv.org/content/10.1101/2023.10.23.23296753v2.supplementary-material) for the specific codes used to create the concept sets. \n",
    "    \n",
    "\n",
    "**03 Extract Demogrphaics.ipynb**: Pull the demographic data and the consent data for all people with WGS in alpha 3. \n",
    "\n",
    "**03 Extract Pathology.ipynb**: Pull the dataset containing participants who have a pathology report. A positive pathology report is used as an exclusion criteria for the screened controls. \n",
    "\n",
    "**03 Extract stool positive.ipynb**: Pull the dataset containing participants who have had a positive stool screening. \n",
    "\n",
    "**03 Extract table 4.1 codes.ipynb**: Pull the dataset containing participants with Ulcerative Colitis or Chron's disease. \n",
    "\n",
    "**03 Extract table 5.2.A codes.ipynb**: Pull the dataset containing participants with CRC noted in their EHR. This defines the largest set of cases, after excluding participants with inherted CRC. \n",
    "\n",
    "**03 Extract table 5.2.B codes.ipynb**: Pull the datasets containing participants with surgical procedures related to CRC. These are used as an inclusion criteria for the cleaned cases. \n",
    "\n",
    "**03 Extract table 5.3.A codes and drugs.ipynb**: Pull the datasets containing participants with chemotherapy or radiation. These are used as an inclusion criteria for the cleaned cases. \n",
    "\n",
    "**03 Extract table 5.3.B codes.ipynb**: Pull the dataset containing participants that have non-CRC cancers. These non-CRC cancers are used as an exclusion criteria for the cleaned cases. See 04 CRC case control for phekb CRC algorithm.ipynb for details. \n",
    "\n",
    "**03 Extract table 6.1 codes.ipynb**: Pull the dataset containing participants with CRC screening procedures.\n",
    "\n",
    "**03 Extract table hereditary CRC cancer participants.ipynb**: Pull the dataset containing the participants who have hereditary CRC noted in their EHR.\n",
    "\n",
    "**04 CRC case control for phekb CRC algorithm.ipynb**: Using the pheKB algorithm (https://phekb.org/phenotype/colorectal-cancer-crc) to identify cases and controls. Extensive cleaning of the concept data sets was performed to ensure that the correct codes were included and incorrect codes were excluded. The following sets of cases and controls are created.\n",
    "\n",
    "    case.set1 := All participants who have 1 or more CRC record, but not inherited. (Called all cases) \n",
    "    case.set3 := case.set1, running pheKB algorithm, NOT removing UC/Chron's (Called cleaned cases)\n",
    "    Screened controls\n",
    "    Unscreened controls. Note: All controls includes screened and unscreened controls\n",
    "\n",
    "\n",
    "**05 Plink commands**: Shell and PLINK commands to combine ancestry informative SNPs in the 1KG, HGDP and All of Us dataset. Then QC and filter using genotypes at all datasets. Then output into separate reference and AoU files to be used later. The AoU data is sepearted out into 10 discrete files due to memory issues.\n",
    "\n",
    "**06 Estimate Admixture using method in PMID: 26482676**: Determine list of related and unrelated subsets to use with calculating admixture estimates and PCs. Calculate admixture using each of the ten All of Us ancestry informative SNP data files, given the unrelated reference data (1KG and HGDP) and the contintental ancestry assignments.\n",
    "\n",
    "**07 Compute PCs of ancestry**: Compute the PCs of ancestry in the unrelated subset of 1KG and HGDP data. Calculate the percent variance explained by the PCs. Project the AoU data onto these PCs of ancestry (ten separate files of participants).\n",
    "\n",
    "**08 Demographics**: Demographics - count, age distribution, sex distribution. Further analysis of ancestral distribution of the OTH cluster. \n",
    "\n",
    "**09 Adjust PRS by PCs or admixture**: Tables with mean and sd of the PRS before and after adjustments. Random sampling from the unscreened controls (Set 1) using distance from the median of (PC1, PC2, PC3, PC4, PC5). Also getting a purely random sample (Set 2) from the unscreened controls for comparison\n",
    "\n",
    "**10 Compare PRS OR**: Tables of OR (beta, C.I.) for 1 SD change in PRS with and without adjusting for age and sex. Figures for OR and C.I. comparing quantiles to the middle. Figures of the observed vs expected upper percentiles, with C.I.. Some analysis in response to reviewers regarding the ADM group. \n",
    "\n",
    "**11 Compare PRS AUC**: Compare the AUC performance of the adjusted PRSs for CRC cleaned case and screened control data with and without adjusting for age and sex. Creating tables with the estimated AUC and bootstrap 95% C.I.s \n",
    "\n",
    "**12 Response to Reviewers**: Further analysis in repsponse to reviewer comments. \n",
    "    \n",
    "    Getting set of highly admixed participants (ADM) and tabulating means/s.d. for the raw and adjusted PRS, and getting figure for the upper tails observed vs expected\n",
    "    Getting mean and s.d. of PC_mu and AD_mu, after scaling for overall variance\n",
    "    Kolmogorov-Smirnov tests. Ho: The distribution is N(0,1). Rejection of these tests indicate the adjusted distribution is not N(0,1) \n",
    "    \n",
    "**13 Response to Reviewers**: Further analysis in response to reviewer comments.\n",
    "\n",
    "    Getting citation for GENESIS package\n",
    "    T-tests for age between cases and controls\n",
    "    Mixed model to test mean difference in PRS, allowing for different variances by ancestry cluster\n",
    "    Logisitc regression to test difference in OR for each adjusted PRS, by ancestry cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
