{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "07 Compute PCs of ancestry\n",
    "\n",
    "Using the list of kept SNPs, calcualte PCs of ancestry in the reference dataset to project onto the AoU data. Also, check that the PCs separate the ancestries appropriately\n",
    "\n",
    "Note: PLINK files were not copied over from the pre-production workspace. The code here if for illustration purposes.\n",
    "\n",
    "Using standard VM 4CPUs, 64 CPUs, 57.6 RAM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Install libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "rm(list=ls())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "\t<li>NULL</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\item NULL\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. NULL\n",
       "2. NULL\n",
       "3. NULL\n",
       "4. NULL\n",
       "5. NULL\n",
       "6. NULL\n",
       "7. NULL\n",
       "8. NULL\n",
       "9. NULL\n",
       "10. NULL\n",
       "11. NULL\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "NULL\n",
       "\n",
       "[[2]]\n",
       "NULL\n",
       "\n",
       "[[3]]\n",
       "NULL\n",
       "\n",
       "[[4]]\n",
       "NULL\n",
       "\n",
       "[[5]]\n",
       "NULL\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "NULL\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n",
       "\n",
       "[[10]]\n",
       "NULL\n",
       "\n",
       "[[11]]\n",
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: lattice\n",
      "\n",
      "Loading required package: survival\n",
      "\n",
      "Loading required package: Formula\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "\n",
      "Attaching package: ‘Hmisc’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    format.pval, units\n",
      "\n",
      "\n",
      "Warning message in system(\"timedatectl\", intern = TRUE):\n",
      "“running command 'timedatectl' had status 1”\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.2 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.8      \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.10\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.1      \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.5.0 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.3      \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.2 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.5      \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msrc()\u001b[39m       masks \u001b[34mHmisc\u001b[39m::src()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m masks \u001b[34mHmisc\u001b[39m::summarize()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "SNPRelate -- supported by Streaming SIMD Extensions 2 (SSE2)\n",
      "\n",
      "Loading required package: foreach\n",
      "\n",
      "\n",
      "Attaching package: ‘foreach’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    accumulate, when\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SeqArray’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stringr’:\n",
      "\n",
      "    fixed\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Hmisc’:\n",
      "\n",
      "    contents\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘GWASTools’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:SeqVarTools’:\n",
      "\n",
      "    alleleFrequency, duplicateDiscordance, getGenotype, iterateFilter,\n",
      "    lastFilter, lastFilter<-, mendelErr, resetIterator\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'bigparallelr'</li><li>'foreach'</li><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SeqArray'</li><li>'bigparallelr'</li><li>'foreach'</li><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SeqVarTools'</li><li>'SeqArray'</li><li>'bigparallelr'</li><li>'foreach'</li><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GWASTools'</li><li>'Biobase'</li><li>'BiocGenerics'</li><li>'SeqVarTools'</li><li>'SeqArray'</li><li>'bigparallelr'</li><li>'foreach'</li><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GWASTools'</li><li>'Biobase'</li><li>'BiocGenerics'</li><li>'SeqVarTools'</li><li>'SeqArray'</li><li>'bigparallelr'</li><li>'foreach'</li><li>'SNPRelate'</li><li>'gdsfmt'</li><li>'GENESIS'</li><li>'forcats'</li><li>'stringr'</li><li>'dplyr'</li><li>'purrr'</li><li>'readr'</li><li>'tidyr'</li><li>'tibble'</li><li>'tidyverse'</li><li>'Hmisc'</li><li>'ggplot2'</li><li>'Formula'</li><li>'survival'</li><li>'lattice'</li><li>'data.table'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'bigparallelr'\n",
       "\\item 'foreach'\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'SeqArray'\n",
       "\\item 'bigparallelr'\n",
       "\\item 'foreach'\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'SeqVarTools'\n",
       "\\item 'SeqArray'\n",
       "\\item 'bigparallelr'\n",
       "\\item 'foreach'\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GWASTools'\n",
       "\\item 'Biobase'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'SeqVarTools'\n",
       "\\item 'SeqArray'\n",
       "\\item 'bigparallelr'\n",
       "\\item 'foreach'\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GWASTools'\n",
       "\\item 'Biobase'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'SeqVarTools'\n",
       "\\item 'SeqArray'\n",
       "\\item 'bigparallelr'\n",
       "\\item 'foreach'\n",
       "\\item 'SNPRelate'\n",
       "\\item 'gdsfmt'\n",
       "\\item 'GENESIS'\n",
       "\\item 'forcats'\n",
       "\\item 'stringr'\n",
       "\\item 'dplyr'\n",
       "\\item 'purrr'\n",
       "\\item 'readr'\n",
       "\\item 'tidyr'\n",
       "\\item 'tibble'\n",
       "\\item 'tidyverse'\n",
       "\\item 'Hmisc'\n",
       "\\item 'ggplot2'\n",
       "\\item 'Formula'\n",
       "\\item 'survival'\n",
       "\\item 'lattice'\n",
       "\\item 'data.table'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. 1. 'data.table'\n",
       "2. 'stats'\n",
       "3. 'graphics'\n",
       "4. 'grDevices'\n",
       "5. 'utils'\n",
       "6. 'datasets'\n",
       "7. 'methods'\n",
       "8. 'base'\n",
       "\n",
       "\n",
       "\n",
       "2. 1. 'Hmisc'\n",
       "2. 'ggplot2'\n",
       "3. 'Formula'\n",
       "4. 'survival'\n",
       "5. 'lattice'\n",
       "6. 'data.table'\n",
       "7. 'stats'\n",
       "8. 'graphics'\n",
       "9. 'grDevices'\n",
       "10. 'utils'\n",
       "11. 'datasets'\n",
       "12. 'methods'\n",
       "13. 'base'\n",
       "\n",
       "\n",
       "\n",
       "3. 1. 'forcats'\n",
       "2. 'stringr'\n",
       "3. 'dplyr'\n",
       "4. 'purrr'\n",
       "5. 'readr'\n",
       "6. 'tidyr'\n",
       "7. 'tibble'\n",
       "8. 'tidyverse'\n",
       "9. 'Hmisc'\n",
       "10. 'ggplot2'\n",
       "11. 'Formula'\n",
       "12. 'survival'\n",
       "13. 'lattice'\n",
       "14. 'data.table'\n",
       "15. 'stats'\n",
       "16. 'graphics'\n",
       "17. 'grDevices'\n",
       "18. 'utils'\n",
       "19. 'datasets'\n",
       "20. 'methods'\n",
       "21. 'base'\n",
       "\n",
       "\n",
       "\n",
       "4. 1. 'GENESIS'\n",
       "2. 'forcats'\n",
       "3. 'stringr'\n",
       "4. 'dplyr'\n",
       "5. 'purrr'\n",
       "6. 'readr'\n",
       "7. 'tidyr'\n",
       "8. 'tibble'\n",
       "9. 'tidyverse'\n",
       "10. 'Hmisc'\n",
       "11. 'ggplot2'\n",
       "12. 'Formula'\n",
       "13. 'survival'\n",
       "14. 'lattice'\n",
       "15. 'data.table'\n",
       "16. 'stats'\n",
       "17. 'graphics'\n",
       "18. 'grDevices'\n",
       "19. 'utils'\n",
       "20. 'datasets'\n",
       "21. 'methods'\n",
       "22. 'base'\n",
       "\n",
       "\n",
       "\n",
       "5. 1. 'gdsfmt'\n",
       "2. 'GENESIS'\n",
       "3. 'forcats'\n",
       "4. 'stringr'\n",
       "5. 'dplyr'\n",
       "6. 'purrr'\n",
       "7. 'readr'\n",
       "8. 'tidyr'\n",
       "9. 'tibble'\n",
       "10. 'tidyverse'\n",
       "11. 'Hmisc'\n",
       "12. 'ggplot2'\n",
       "13. 'Formula'\n",
       "14. 'survival'\n",
       "15. 'lattice'\n",
       "16. 'data.table'\n",
       "17. 'stats'\n",
       "18. 'graphics'\n",
       "19. 'grDevices'\n",
       "20. 'utils'\n",
       "21. 'datasets'\n",
       "22. 'methods'\n",
       "23. 'base'\n",
       "\n",
       "\n",
       "\n",
       "6. 1. 'SNPRelate'\n",
       "2. 'gdsfmt'\n",
       "3. 'GENESIS'\n",
       "4. 'forcats'\n",
       "5. 'stringr'\n",
       "6. 'dplyr'\n",
       "7. 'purrr'\n",
       "8. 'readr'\n",
       "9. 'tidyr'\n",
       "10. 'tibble'\n",
       "11. 'tidyverse'\n",
       "12. 'Hmisc'\n",
       "13. 'ggplot2'\n",
       "14. 'Formula'\n",
       "15. 'survival'\n",
       "16. 'lattice'\n",
       "17. 'data.table'\n",
       "18. 'stats'\n",
       "19. 'graphics'\n",
       "20. 'grDevices'\n",
       "21. 'utils'\n",
       "22. 'datasets'\n",
       "23. 'methods'\n",
       "24. 'base'\n",
       "\n",
       "\n",
       "\n",
       "7. 1. 'bigparallelr'\n",
       "2. 'foreach'\n",
       "3. 'SNPRelate'\n",
       "4. 'gdsfmt'\n",
       "5. 'GENESIS'\n",
       "6. 'forcats'\n",
       "7. 'stringr'\n",
       "8. 'dplyr'\n",
       "9. 'purrr'\n",
       "10. 'readr'\n",
       "11. 'tidyr'\n",
       "12. 'tibble'\n",
       "13. 'tidyverse'\n",
       "14. 'Hmisc'\n",
       "15. 'ggplot2'\n",
       "16. 'Formula'\n",
       "17. 'survival'\n",
       "18. 'lattice'\n",
       "19. 'data.table'\n",
       "20. 'stats'\n",
       "21. 'graphics'\n",
       "22. 'grDevices'\n",
       "23. 'utils'\n",
       "24. 'datasets'\n",
       "25. 'methods'\n",
       "26. 'base'\n",
       "\n",
       "\n",
       "\n",
       "8. 1. 'SeqArray'\n",
       "2. 'bigparallelr'\n",
       "3. 'foreach'\n",
       "4. 'SNPRelate'\n",
       "5. 'gdsfmt'\n",
       "6. 'GENESIS'\n",
       "7. 'forcats'\n",
       "8. 'stringr'\n",
       "9. 'dplyr'\n",
       "10. 'purrr'\n",
       "11. 'readr'\n",
       "12. 'tidyr'\n",
       "13. 'tibble'\n",
       "14. 'tidyverse'\n",
       "15. 'Hmisc'\n",
       "16. 'ggplot2'\n",
       "17. 'Formula'\n",
       "18. 'survival'\n",
       "19. 'lattice'\n",
       "20. 'data.table'\n",
       "21. 'stats'\n",
       "22. 'graphics'\n",
       "23. 'grDevices'\n",
       "24. 'utils'\n",
       "25. 'datasets'\n",
       "26. 'methods'\n",
       "27. 'base'\n",
       "\n",
       "\n",
       "\n",
       "9. 1. 'SeqVarTools'\n",
       "2. 'SeqArray'\n",
       "3. 'bigparallelr'\n",
       "4. 'foreach'\n",
       "5. 'SNPRelate'\n",
       "6. 'gdsfmt'\n",
       "7. 'GENESIS'\n",
       "8. 'forcats'\n",
       "9. 'stringr'\n",
       "10. 'dplyr'\n",
       "11. 'purrr'\n",
       "12. 'readr'\n",
       "13. 'tidyr'\n",
       "14. 'tibble'\n",
       "15. 'tidyverse'\n",
       "16. 'Hmisc'\n",
       "17. 'ggplot2'\n",
       "18. 'Formula'\n",
       "19. 'survival'\n",
       "20. 'lattice'\n",
       "21. 'data.table'\n",
       "22. 'stats'\n",
       "23. 'graphics'\n",
       "24. 'grDevices'\n",
       "25. 'utils'\n",
       "26. 'datasets'\n",
       "27. 'methods'\n",
       "28. 'base'\n",
       "\n",
       "\n",
       "\n",
       "10. 1. 'GWASTools'\n",
       "2. 'Biobase'\n",
       "3. 'BiocGenerics'\n",
       "4. 'SeqVarTools'\n",
       "5. 'SeqArray'\n",
       "6. 'bigparallelr'\n",
       "7. 'foreach'\n",
       "8. 'SNPRelate'\n",
       "9. 'gdsfmt'\n",
       "10. 'GENESIS'\n",
       "11. 'forcats'\n",
       "12. 'stringr'\n",
       "13. 'dplyr'\n",
       "14. 'purrr'\n",
       "15. 'readr'\n",
       "16. 'tidyr'\n",
       "17. 'tibble'\n",
       "18. 'tidyverse'\n",
       "19. 'Hmisc'\n",
       "20. 'ggplot2'\n",
       "21. 'Formula'\n",
       "22. 'survival'\n",
       "23. 'lattice'\n",
       "24. 'data.table'\n",
       "25. 'stats'\n",
       "26. 'graphics'\n",
       "27. 'grDevices'\n",
       "28. 'utils'\n",
       "29. 'datasets'\n",
       "30. 'methods'\n",
       "31. 'base'\n",
       "\n",
       "\n",
       "\n",
       "11. 1. 'GWASTools'\n",
       "2. 'Biobase'\n",
       "3. 'BiocGenerics'\n",
       "4. 'SeqVarTools'\n",
       "5. 'SeqArray'\n",
       "6. 'bigparallelr'\n",
       "7. 'foreach'\n",
       "8. 'SNPRelate'\n",
       "9. 'gdsfmt'\n",
       "10. 'GENESIS'\n",
       "11. 'forcats'\n",
       "12. 'stringr'\n",
       "13. 'dplyr'\n",
       "14. 'purrr'\n",
       "15. 'readr'\n",
       "16. 'tidyr'\n",
       "17. 'tibble'\n",
       "18. 'tidyverse'\n",
       "19. 'Hmisc'\n",
       "20. 'ggplot2'\n",
       "21. 'Formula'\n",
       "22. 'survival'\n",
       "23. 'lattice'\n",
       "24. 'data.table'\n",
       "25. 'stats'\n",
       "26. 'graphics'\n",
       "27. 'grDevices'\n",
       "28. 'utils'\n",
       "29. 'datasets'\n",
       "30. 'methods'\n",
       "31. 'base'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"data.table\" \"stats\"      \"graphics\"   \"grDevices\"  \"utils\"     \n",
       "[6] \"datasets\"   \"methods\"    \"base\"      \n",
       "\n",
       "[[2]]\n",
       " [1] \"Hmisc\"      \"ggplot2\"    \"Formula\"    \"survival\"   \"lattice\"   \n",
       " [6] \"data.table\" \"stats\"      \"graphics\"   \"grDevices\"  \"utils\"     \n",
       "[11] \"datasets\"   \"methods\"    \"base\"      \n",
       "\n",
       "[[3]]\n",
       " [1] \"forcats\"    \"stringr\"    \"dplyr\"      \"purrr\"      \"readr\"     \n",
       " [6] \"tidyr\"      \"tibble\"     \"tidyverse\"  \"Hmisc\"      \"ggplot2\"   \n",
       "[11] \"Formula\"    \"survival\"   \"lattice\"    \"data.table\" \"stats\"     \n",
       "[16] \"graphics\"   \"grDevices\"  \"utils\"      \"datasets\"   \"methods\"   \n",
       "[21] \"base\"      \n",
       "\n",
       "[[4]]\n",
       " [1] \"GENESIS\"    \"forcats\"    \"stringr\"    \"dplyr\"      \"purrr\"     \n",
       " [6] \"readr\"      \"tidyr\"      \"tibble\"     \"tidyverse\"  \"Hmisc\"     \n",
       "[11] \"ggplot2\"    \"Formula\"    \"survival\"   \"lattice\"    \"data.table\"\n",
       "[16] \"stats\"      \"graphics\"   \"grDevices\"  \"utils\"      \"datasets\"  \n",
       "[21] \"methods\"    \"base\"      \n",
       "\n",
       "[[5]]\n",
       " [1] \"gdsfmt\"     \"GENESIS\"    \"forcats\"    \"stringr\"    \"dplyr\"     \n",
       " [6] \"purrr\"      \"readr\"      \"tidyr\"      \"tibble\"     \"tidyverse\" \n",
       "[11] \"Hmisc\"      \"ggplot2\"    \"Formula\"    \"survival\"   \"lattice\"   \n",
       "[16] \"data.table\" \"stats\"      \"graphics\"   \"grDevices\"  \"utils\"     \n",
       "[21] \"datasets\"   \"methods\"    \"base\"      \n",
       "\n",
       "[[6]]\n",
       " [1] \"SNPRelate\"  \"gdsfmt\"     \"GENESIS\"    \"forcats\"    \"stringr\"   \n",
       " [6] \"dplyr\"      \"purrr\"      \"readr\"      \"tidyr\"      \"tibble\"    \n",
       "[11] \"tidyverse\"  \"Hmisc\"      \"ggplot2\"    \"Formula\"    \"survival\"  \n",
       "[16] \"lattice\"    \"data.table\" \"stats\"      \"graphics\"   \"grDevices\" \n",
       "[21] \"utils\"      \"datasets\"   \"methods\"    \"base\"      \n",
       "\n",
       "[[7]]\n",
       " [1] \"bigparallelr\" \"foreach\"      \"SNPRelate\"    \"gdsfmt\"       \"GENESIS\"     \n",
       " [6] \"forcats\"      \"stringr\"      \"dplyr\"        \"purrr\"        \"readr\"       \n",
       "[11] \"tidyr\"        \"tibble\"       \"tidyverse\"    \"Hmisc\"        \"ggplot2\"     \n",
       "[16] \"Formula\"      \"survival\"     \"lattice\"      \"data.table\"   \"stats\"       \n",
       "[21] \"graphics\"     \"grDevices\"    \"utils\"        \"datasets\"     \"methods\"     \n",
       "[26] \"base\"        \n",
       "\n",
       "[[8]]\n",
       " [1] \"SeqArray\"     \"bigparallelr\" \"foreach\"      \"SNPRelate\"    \"gdsfmt\"      \n",
       " [6] \"GENESIS\"      \"forcats\"      \"stringr\"      \"dplyr\"        \"purrr\"       \n",
       "[11] \"readr\"        \"tidyr\"        \"tibble\"       \"tidyverse\"    \"Hmisc\"       \n",
       "[16] \"ggplot2\"      \"Formula\"      \"survival\"     \"lattice\"      \"data.table\"  \n",
       "[21] \"stats\"        \"graphics\"     \"grDevices\"    \"utils\"        \"datasets\"    \n",
       "[26] \"methods\"      \"base\"        \n",
       "\n",
       "[[9]]\n",
       " [1] \"SeqVarTools\"  \"SeqArray\"     \"bigparallelr\" \"foreach\"      \"SNPRelate\"   \n",
       " [6] \"gdsfmt\"       \"GENESIS\"      \"forcats\"      \"stringr\"      \"dplyr\"       \n",
       "[11] \"purrr\"        \"readr\"        \"tidyr\"        \"tibble\"       \"tidyverse\"   \n",
       "[16] \"Hmisc\"        \"ggplot2\"      \"Formula\"      \"survival\"     \"lattice\"     \n",
       "[21] \"data.table\"   \"stats\"        \"graphics\"     \"grDevices\"    \"utils\"       \n",
       "[26] \"datasets\"     \"methods\"      \"base\"        \n",
       "\n",
       "[[10]]\n",
       " [1] \"GWASTools\"    \"Biobase\"      \"BiocGenerics\" \"SeqVarTools\"  \"SeqArray\"    \n",
       " [6] \"bigparallelr\" \"foreach\"      \"SNPRelate\"    \"gdsfmt\"       \"GENESIS\"     \n",
       "[11] \"forcats\"      \"stringr\"      \"dplyr\"        \"purrr\"        \"readr\"       \n",
       "[16] \"tidyr\"        \"tibble\"       \"tidyverse\"    \"Hmisc\"        \"ggplot2\"     \n",
       "[21] \"Formula\"      \"survival\"     \"lattice\"      \"data.table\"   \"stats\"       \n",
       "[26] \"graphics\"     \"grDevices\"    \"utils\"        \"datasets\"     \"methods\"     \n",
       "[31] \"base\"        \n",
       "\n",
       "[[11]]\n",
       " [1] \"GWASTools\"    \"Biobase\"      \"BiocGenerics\" \"SeqVarTools\"  \"SeqArray\"    \n",
       " [6] \"bigparallelr\" \"foreach\"      \"SNPRelate\"    \"gdsfmt\"       \"GENESIS\"     \n",
       "[11] \"forcats\"      \"stringr\"      \"dplyr\"        \"purrr\"        \"readr\"       \n",
       "[16] \"tidyr\"        \"tibble\"       \"tidyverse\"    \"Hmisc\"        \"ggplot2\"     \n",
       "[21] \"Formula\"      \"survival\"     \"lattice\"      \"data.table\"   \"stats\"       \n",
       "[26] \"graphics\"     \"grDevices\"    \"utils\"        \"datasets\"     \"methods\"     \n",
       "[31] \"base\"        \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "my.packages <- c('data.table','Hmisc','tidyverse','GENESIS','gdsfmt','SNPRelate','bigparallelr','SeqArray',\n",
    "                 'SeqVarTools','GWASTools','ggplot2')\n",
    "lapply(my.packages,\n",
    "       function(pkg) { if(! pkg %in% installed.packages()) { install.packages(pkg)} } )\n",
    "lapply(my.packages,library,character.only = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.2.2 Patched (2022-11-10 r83330)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] GWASTools_1.44.0    Biobase_2.58.0      BiocGenerics_0.44.0\n",
       " [4] SeqVarTools_1.36.0  SeqArray_1.38.0     bigparallelr_0.3.2 \n",
       " [7] foreach_1.5.2       SNPRelate_1.32.1    gdsfmt_1.34.0      \n",
       "[10] GENESIS_2.28.0      forcats_0.5.2       stringr_1.5.0      \n",
       "[13] dplyr_1.0.10        purrr_0.3.5         readr_2.1.3        \n",
       "[16] tidyr_1.2.1         tibble_3.1.8        tidyverse_1.3.2    \n",
       "[19] Hmisc_4.7-2         ggplot2_3.4.0       Formula_1.2-4      \n",
       "[22] survival_3.4-0      lattice_0.20-45     data.table_1.14.6  \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] readxl_1.4.1           uuid_1.1-0             backports_1.4.1       \n",
       "  [4] repr_1.1.4             splines_4.2.2          operator.tools_1.6.3  \n",
       "  [7] BiocParallel_1.32.4    GenomeInfoDb_1.34.4    digest_0.6.30         \n",
       " [10] htmltools_0.5.4        fansi_1.0.3            magrittr_2.0.3        \n",
       " [13] checkmate_2.1.0        memoise_2.0.1          doParallel_1.0.17     \n",
       " [16] googlesheets4_1.0.1    cluster_2.1.4          bigassertr_0.1.6      \n",
       " [19] tzdb_0.3.0             Biostrings_2.66.0      modelr_0.1.10         \n",
       " [22] formula.tools_1.7.1    sandwich_3.0-2         timechange_0.1.1      \n",
       " [25] jpeg_0.1-10            colorspace_2.0-3       blob_1.2.3            \n",
       " [28] rvest_1.0.3            haven_2.5.1            xfun_0.35             \n",
       " [31] crayon_1.5.2           RCurl_1.98-1.9         jsonlite_1.8.4        \n",
       " [34] flock_0.7              zoo_1.8-11             iterators_1.0.14      \n",
       " [37] glue_1.6.2             gtable_0.3.1           gargle_1.2.1          \n",
       " [40] zlibbioc_1.44.0        XVector_0.38.0         MatrixModels_0.5-1    \n",
       " [43] SparseM_1.81           scales_1.2.1           DBI_1.1.3             \n",
       " [46] Rcpp_1.0.9             htmlTable_2.4.1        quantsmooth_1.64.0    \n",
       " [49] foreign_0.8-84         bit_4.0.5              stats4_4.2.2          \n",
       " [52] htmlwidgets_1.5.4      httr_1.4.4             logistf_1.24.1        \n",
       " [55] RColorBrewer_1.1-3     ellipsis_0.3.2         mice_3.15.0           \n",
       " [58] pkgconfig_2.0.3        nnet_7.3-18            dbplyr_2.2.1          \n",
       " [61] deldir_1.0-6           utf8_1.2.2             DNAcopy_1.72.2        \n",
       " [64] tidyselect_1.2.0       rlang_1.0.6            munsell_0.5.0         \n",
       " [67] cellranger_1.1.0       tools_4.2.2            cachem_1.0.6          \n",
       " [70] cli_3.4.1              generics_0.1.3         RSQLite_2.2.19        \n",
       " [73] broom_1.0.1            evaluate_0.18          fastmap_1.1.0         \n",
       " [76] knitr_1.41             bit64_4.0.5            fs_1.5.2              \n",
       " [79] nlme_3.1-160           GWASExactHW_1.01       quantreg_5.94         \n",
       " [82] xml2_1.3.3             compiler_4.2.2         rstudioapi_0.14       \n",
       " [85] png_0.1-8              reprex_2.0.2           stringi_1.7.8         \n",
       " [88] IRdisplay_1.1          Matrix_1.5-3           vctrs_0.5.1           \n",
       " [91] pillar_1.8.1           lifecycle_1.0.3        lmtest_0.9-40         \n",
       " [94] bitops_1.0-7           GenomicRanges_1.50.1   R6_2.5.1              \n",
       " [97] latticeExtra_0.6-30    gridExtra_2.3          IRanges_2.32.0        \n",
       "[100] codetools_0.2-18       MASS_7.3-58.1          assertthat_0.2.1      \n",
       "[103] withr_2.5.0            S4Vectors_0.36.1       GenomeInfoDbData_1.2.9\n",
       "[106] mgcv_1.8-41            parallel_4.2.2         hms_1.1.2             \n",
       "[109] grid_4.2.2             rpart_4.1.19           IRkernel_1.3.1        \n",
       "[112] googledrive_2.0.0      pbdZMQ_0.3-8           lubridate_1.9.0       \n",
       "[115] base64enc_0.1-3        interp_1.1-3          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "63"
      ],
      "text/latex": [
       "63"
      ],
      "text/markdown": [
       "63"
      ],
      "text/plain": [
       "[1] 63"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "(num.threads <- nb_cores())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Reference data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the plink file data\n",
    "cp.command <- \"gsutil -m cp -r gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/all_combined_PCSNPS_final_ref.* .\"\n",
    "system(cp.command)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed.fn <- \"all_combined_PCSNPS_final_ref.bed\"\n",
    "fam.fn <- \"all_combined_PCSNPS_final_ref.fam\"\n",
    "bim.fn <- \"all_combined_PCSNPS_final_ref.bim\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_ref.bed'\n",
      "        SNP-major mode (Sample X SNP), 297.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_ref.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_ref.bim'\n",
      "Mon Feb 13 16:41:17 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 4151 samples, 300387 SNPs ...\n",
      "[==================================================] 100%, completed, 3s  \n",
      "Mon Feb 13 16:41:23 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'ref-pcsnps.gds' (299.0M)\n",
      "    # of fragments: 39\n",
      "    save to 'ref-pcsnps.gds.tmp'\n",
      "    rename 'ref-pcsnps.gds.tmp' (299.0M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "The file name: /home/jupyter/workspaces/crcprsanalysisinteractivenotebooks/ref-pcsnps.gds \n",
      "The total number of samples: 4151 \n",
      "The total number of SNPs: 300387 \n",
      "SNP genotypes are stored in SNP-major mode (Sample X SNP).\n"
     ]
    }
   ],
   "source": [
    "# create the gds file\n",
    "snpgdsBED2GDS(bed.fn, fam.fn, bim.fn, \"ref-pcsnps.gds\")\n",
    "snpgdsSummary(\"ref-pcsnps.gds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load in list of related and unrelated reference samples previously calculated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in system(cp.command, intern = TRUE):\n",
      "“running command 'gsutil -m cp -r gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/reference/data/PCSNPSref-sampset.RData .' had status 1”\n"
     ]
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)\n",
       "attr(,\"status\")\n",
       "[1] 1\n",
       "attr(,\"errmsg\")\n",
       "[1] \"Resource temporarily unavailable\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "infile <- \"ref-sampset.RData\"\n",
    "cp.command <- paste0(\"gsutil -m cp -r gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/reference/data/PCSNPS\",infile,\" .\")\n",
    "system(cp.command, intern=TRUE)\n",
    "load(infile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>rels</dt><dd>711</dd><dt>unrels</dt><dd>3440</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[rels] 711\n",
       "\\item[unrels] 3440\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "rels\n",
       ":   711unrels\n",
       ":   3440\n",
       "\n"
      ],
      "text/plain": [
       "  rels unrels \n",
       "   711   3440 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sapply(sampset1, length)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate PCA using unrelated reference samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "genofile <- snpgdsOpen(\"ref-pcsnps.gds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Principal Component Analysis (PCA) on genotypes:\n",
      "Excluding 0 SNP on non-autosomes\n",
      "Excluding 0 SNP (monomorphic: TRUE, MAF: NaN, missing rate: NaN)\n",
      "    # of samples: 3,440\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    # of principal components: 32\n",
      "PCA:    the sum of all selected genotypes (0,1,2) = 250283131\n",
      "CPU capabilities: Double-Precision SSE2\n",
      "Mon Feb 13 16:44:17 2023    (internal increment: 1700)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 35s\n",
      "Mon Feb 13 16:44:52 2023    Begin (eigenvalues and eigenvectors)\n",
      "Mon Feb 13 16:45:01 2023    Done.\n"
     ]
    }
   ],
   "source": [
    "pca.ref <- snpgdsPCA(genofile,sample.id = sampset1$unrels,num.thread=num.threads) # took 13 minutes with 4 CPUs, 15 GB RAM.\n",
    "#takes less than a minute with 64 CPUs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 8\n",
      " $ sample.id: chr [1:3440] \"CHMI_CHMI3_WGS2\" \"HG00096\" \"HG00097\" \"HG00099\" ...\n",
      " $ snp.id   : chr [1:300387] \"chr1:819049:T:G\" \"chr1:833994:G:A\" \"chr1:838618:C:T\" \"chr1:886546:C:G\" ...\n",
      " $ eigenval : num [1:3440] 184.66 82.74 28.36 21.11 9.25 ...\n",
      " $ eigenvect: num [1:3440, 1:32] -0.00954 -0.01021 -0.01026 -0.01024 -0.01014 ...\n",
      " $ varprop  : num [1:3440] 0.0537 0.02406 0.00825 0.00614 0.00269 ...\n",
      " $ TraceXTX : num 2.19e+09\n",
      " $ Bayesian : logi FALSE\n",
      " $ genmat   : NULL\n",
      " - attr(*, \"class\")= chr \"snpgdsPCAClass\"\n"
     ]
    }
   ],
   "source": [
    "str(pca.ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>5.37</li><li>2.41</li><li>0.82</li><li>0.61</li><li>0.27</li><li>0.23</li><li>0.17</li><li>0.16</li><li>0.14</li><li>0.12</li><li>0.1</li><li>0.1</li><li>0.09</li><li>0.07</li><li>0.07</li><li>0.07</li><li>0.06</li><li>0.06</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li><li>0.05</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 5.37\n",
       "\\item 2.41\n",
       "\\item 0.82\n",
       "\\item 0.61\n",
       "\\item 0.27\n",
       "\\item 0.23\n",
       "\\item 0.17\n",
       "\\item 0.16\n",
       "\\item 0.14\n",
       "\\item 0.12\n",
       "\\item 0.1\n",
       "\\item 0.1\n",
       "\\item 0.09\n",
       "\\item 0.07\n",
       "\\item 0.07\n",
       "\\item 0.07\n",
       "\\item 0.06\n",
       "\\item 0.06\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\item 0.05\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 5.37\n",
       "2. 2.41\n",
       "3. 0.82\n",
       "4. 0.61\n",
       "5. 0.27\n",
       "6. 0.23\n",
       "7. 0.17\n",
       "8. 0.16\n",
       "9. 0.14\n",
       "10. 0.12\n",
       "11. 0.1\n",
       "12. 0.1\n",
       "13. 0.09\n",
       "14. 0.07\n",
       "15. 0.07\n",
       "16. 0.07\n",
       "17. 0.06\n",
       "18. 0.06\n",
       "19. 0.05\n",
       "20. 0.05\n",
       "21. 0.05\n",
       "22. 0.05\n",
       "23. 0.05\n",
       "24. 0.05\n",
       "25. 0.05\n",
       "26. 0.05\n",
       "27. 0.05\n",
       "28. 0.05\n",
       "29. 0.05\n",
       "30. 0.05\n",
       "31. 0.05\n",
       "32. 0.05\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] 5.37 2.41 0.82 0.61 0.27 0.23 0.17 0.16 0.14 0.12 0.10 0.10 0.09 0.07 0.07\n",
       "[16] 0.07 0.06 0.06 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05\n",
       "[31] 0.05 0.05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 32 of type dbl</caption>\n",
       "<tbody>\n",
       "\t<tr><td>1.00</td><td>2.00</td><td>3.00</td><td>4.00</td><td>5.00</td><td>6.00</td><td>7.00</td><td>8.00</td><td>9.00</td><td>10.00</td><td>⋯</td><td>23.00</td><td>24.00</td><td>25.00</td><td>26.00</td><td>27.00</td><td>28.00</td><td>29.00</td><td>30.00</td><td>31.00</td><td>32.00</td></tr>\n",
       "\t<tr><td>5.37</td><td>2.41</td><td>0.82</td><td>0.61</td><td>0.27</td><td>0.23</td><td>0.17</td><td>0.16</td><td>0.14</td><td> 0.12</td><td>⋯</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td><td> 0.05</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 32 of type dbl\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       "\t 1.00 & 2.00 & 3.00 & 4.00 & 5.00 & 6.00 & 7.00 & 8.00 & 9.00 & 10.00 & ⋯ & 23.00 & 24.00 & 25.00 & 26.00 & 27.00 & 28.00 & 29.00 & 30.00 & 31.00 & 32.00\\\\\n",
       "\t 5.37 & 2.41 & 0.82 & 0.61 & 0.27 & 0.23 & 0.17 & 0.16 & 0.14 &  0.12 & ⋯ &  0.05 &  0.05 &  0.05 &  0.05 &  0.05 &  0.05 &  0.05 &  0.05 &  0.05 &  0.05\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 32 of type dbl\n",
       "\n",
       "| 1.00 | 2.00 | 3.00 | 4.00 | 5.00 | 6.00 | 7.00 | 8.00 | 9.00 | 10.00 | ⋯ | 23.00 | 24.00 | 25.00 | 26.00 | 27.00 | 28.00 | 29.00 | 30.00 | 31.00 | 32.00 |\n",
       "| 5.37 | 2.41 | 0.82 | 0.61 | 0.27 | 0.23 | 0.17 | 0.16 | 0.14 |  0.12 | ⋯ |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |  0.05 |\n",
       "\n"
      ],
      "text/plain": [
       "     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]\n",
       "[1,] 1.00 2.00 3.00 4.00 5.00 6.00 7.00 8.00 9.00 10.00 ⋯     23.00 24.00 25.00\n",
       "[2,] 5.37 2.41 0.82 0.61 0.27 0.23 0.17 0.16 0.14  0.12 ⋯      0.05  0.05  0.05\n",
       "     [,15] [,16] [,17] [,18] [,19] [,20] [,21]\n",
       "[1,] 26.00 27.00 28.00 29.00 30.00 31.00 32.00\n",
       "[2,]  0.05  0.05  0.05  0.05  0.05  0.05  0.05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pc.percent <- pca.ref$varprop*100\n",
    "head(round(pc.percent, 2),n=32L)\n",
    "rbind(c(1L:32L),head(round(pc.percent, 2),n=32L))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP Loading:\n",
      "    # of samples: 3,440\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "SNP Loading:    the sum of all selected genotypes (0,1,2) = 250283131\n",
      "Mon Feb 13 16:45:38 2023    (internal increment: 13628)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 2s\n",
      "Mon Feb 13 16:45:40 2023    Done.\n"
     ]
    }
   ],
   "source": [
    "# only want to do this once. Get Loadings based on the reference samples\n",
    "snp.load <- snpgdsPCASNPLoading(pca.ref, gdsobj=genofile,num.thread=num.threads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample Loading:\n",
      "    # of samples: 4,151\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 303551849\n",
      "Mon Feb 13 16:49:29 2023    (internal increment: 11296)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 3s\n",
      "Mon Feb 13 16:49:32 2023    Done.\n"
     ]
    }
   ],
   "source": [
    "#project onto all reference samples\n",
    "samp.load <- snpgdsPCASampLoading(snp.load, gdsobj=genofile,num.thread=num.threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output the PCs for the reference samples\n",
    "#pcs <- all.pcs <- pca.ref$eigenvect\n",
    "#ref.sample.id <- all.ids <- pca.ref$sample.id\n",
    "pcs <- all.pcs <- samp.load$eigenvect\n",
    "ref.sample.id <- all.ids <- samp.load$sample.id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project PCA onto AoU participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou00.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou00.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou00.bim'\n",
      "Mon Feb 13 16:51:53 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:52:03 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 717467006\n",
      "Mon Feb 13 16:52:09 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:52:15 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou01.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou01.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou01.bim'\n",
      "Mon Feb 13 16:52:33 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 4s\n",
      "Mon Feb 13 16:52:43 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 719135994\n",
      "Mon Feb 13 16:52:48 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:52:54 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou02.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou02.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou02.bim'\n",
      "Mon Feb 13 16:53:13 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 4s\n",
      "Mon Feb 13 16:53:20 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 718070438\n",
      "Mon Feb 13 16:53:25 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:53:31 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou03.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou03.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou03.bim'\n",
      "Mon Feb 13 16:53:49 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:53:58 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 719367096\n",
      "Mon Feb 13 16:54:03 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:54:09 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou04.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou04.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou04.bim'\n",
      "Mon Feb 13 16:54:27 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:54:35 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 718315220\n",
      "Mon Feb 13 16:54:40 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:54:46 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou05.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou05.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou05.bim'\n",
      "Mon Feb 13 16:55:05 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:55:12 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 718415074\n",
      "Mon Feb 13 16:55:18 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:55:24 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou06.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou06.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou06.bim'\n",
      "Mon Feb 13 16:55:42 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 4s\n",
      "Mon Feb 13 16:55:49 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 707433222\n",
      "Mon Feb 13 16:55:55 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:56:01 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou07.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou07.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou07.bim'\n",
      "Mon Feb 13 16:56:19 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:56:27 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 700194804\n",
      "Mon Feb 13 16:56:32 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:56:38 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou08.bed'\n",
      "        SNP-major mode (Sample X SNP), 706.4M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou08.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou08.bim'\n",
      "Mon Feb 13 16:56:57 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9863 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 5s\n",
      "Mon Feb 13 16:57:05 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (708.1M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (708.1M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,863\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 698996465\n",
      "Mon Feb 13 16:57:10 2023    (internal increment: 4752)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:57:16 2023    Done.\n",
      "Start file conversion from PLINK BED to SNP GDS ...\n",
      "    BED file: 'all_combined_PCSNPS_final_aou09.bed'\n",
      "        SNP-major mode (Sample X SNP), 705.9M\n",
      "    FAM file: 'all_combined_PCSNPS_final_aou09.fam'\n",
      "    BIM file: 'all_combined_PCSNPS_final_aou09.bim'\n",
      "Mon Feb 13 16:57:35 2023     (store sample id, snp id, position, and chromosome)\n",
      "    start writing: 9855 samples, 300387 SNPs ...\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 4s\n",
      "Mon Feb 13 16:57:42 2023 \tDone.\n",
      "Optimize the access efficiency ...\n",
      "Clean up the fragments of GDS file:\n",
      "    open the file 'aou-pcsnps.gds' (707.5M)\n",
      "    # of fragments: 39\n",
      "    save to 'aou-pcsnps.gds.tmp'\n",
      "    rename 'aou-pcsnps.gds.tmp' (707.5M, reduced: 252B)\n",
      "    # of fragments: 18\n",
      "Sample Loading:\n",
      "    # of samples: 9,855\n",
      "    # of SNPs: 300,387\n",
      "    using 63 threads\n",
      "    using the top 32 eigenvectors\n",
      "Sample Loading:    the sum of all selected genotypes (0,1,2) = 691257710\n",
      "Mon Feb 13 16:57:47 2023    (internal increment: 4756)\n",
      "\r",
      "[..................................................]  0%, ETC: ---        \r",
      "[==================================================] 100%, completed, 6s\n",
      "Mon Feb 13 16:57:53 2023    Done.\n"
     ]
    }
   ],
   "source": [
    "# for loop applied to all of the AoU PLINK files\n",
    "for(i in 0:9)\n",
    "{\n",
    "    # Get the AoU file \n",
    "    (infile <- paste0(\"all_combined_PCSNPS_final_aou0\",i))\n",
    "    (cp.command <- paste0(\"gsutil -m cp -r gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/\",infile,\".* .\"))\n",
    "    system(cp.command)\n",
    "    \n",
    "    #set up for the GDS\n",
    "    (bed.aou.fn <- paste0(\"all_combined_PCSNPS_final_aou0\",i,\".bed\"))\n",
    "    (fam.aou.fn <- paste0(\"all_combined_PCSNPS_final_aou0\",i,\".fam\"))\n",
    "    (bim.aou.fn <- paste0(\"all_combined_PCSNPS_final_aou0\",i,\".bim\"))\n",
    "    snpgdsBED2GDS(bed.aou.fn, fam.aou.fn, bim.aou.fn, \"aou-pcsnps.gds\")\n",
    "    genofile.aou <- snpgdsOpen(\"aou-pcsnps.gds\")\n",
    "\n",
    "    #Project onto the AoU subset\n",
    "    samp.load <- snpgdsPCASampLoading(snp.load, gdsobj=genofile.aou,num.thread=num.threads)\n",
    "\n",
    "    # get output\n",
    "    sample.id <- samp.load$sample.id\n",
    "    pcs <- samp.load$eigenvect\n",
    "    all.pcs <- rbind(all.pcs,pcs) #If I can output a single file, this would be best\n",
    "    all.ids <- c(all.ids, sample.id)\n",
    "\n",
    "    #close the connection to the AoU data\n",
    "    snpgdsClose(genofile.aou)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Output the PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " chr [1:102773, 1:33] \"CHMI_CHMI3_WGS2\" \"HG00096\" \"HG00097\" \"HG00099\" ...\n",
      " - attr(*, \"dimnames\")=List of 2\n",
      "  ..$ : NULL\n",
      "  ..$ : chr [1:33] \"all.ids\" \"\" \"\" \"\" ...\n"
     ]
    }
   ],
   "source": [
    "# output the PCs\n",
    "output <- cbind(all.ids, all.pcs)\n",
    "str(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'all.ids'</li><li>'V2'</li><li>'V3'</li><li>'V4'</li><li>'V5'</li><li>'V6'</li><li>'V7'</li><li>'V8'</li><li>'V9'</li><li>'V10'</li><li>'V11'</li><li>'V12'</li><li>'V13'</li><li>'V14'</li><li>'V15'</li><li>'V16'</li><li>'V17'</li><li>'V18'</li><li>'V19'</li><li>'V20'</li><li>'V21'</li><li>'V22'</li><li>'V23'</li><li>'V24'</li><li>'V25'</li><li>'V26'</li><li>'V27'</li><li>'V28'</li><li>'V29'</li><li>'V30'</li><li>'V31'</li><li>'V32'</li><li>'V33'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'all.ids'\n",
       "\\item 'V2'\n",
       "\\item 'V3'\n",
       "\\item 'V4'\n",
       "\\item 'V5'\n",
       "\\item 'V6'\n",
       "\\item 'V7'\n",
       "\\item 'V8'\n",
       "\\item 'V9'\n",
       "\\item 'V10'\n",
       "\\item 'V11'\n",
       "\\item 'V12'\n",
       "\\item 'V13'\n",
       "\\item 'V14'\n",
       "\\item 'V15'\n",
       "\\item 'V16'\n",
       "\\item 'V17'\n",
       "\\item 'V18'\n",
       "\\item 'V19'\n",
       "\\item 'V20'\n",
       "\\item 'V21'\n",
       "\\item 'V22'\n",
       "\\item 'V23'\n",
       "\\item 'V24'\n",
       "\\item 'V25'\n",
       "\\item 'V26'\n",
       "\\item 'V27'\n",
       "\\item 'V28'\n",
       "\\item 'V29'\n",
       "\\item 'V30'\n",
       "\\item 'V31'\n",
       "\\item 'V32'\n",
       "\\item 'V33'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'all.ids'\n",
       "2. 'V2'\n",
       "3. 'V3'\n",
       "4. 'V4'\n",
       "5. 'V5'\n",
       "6. 'V6'\n",
       "7. 'V7'\n",
       "8. 'V8'\n",
       "9. 'V9'\n",
       "10. 'V10'\n",
       "11. 'V11'\n",
       "12. 'V12'\n",
       "13. 'V13'\n",
       "14. 'V14'\n",
       "15. 'V15'\n",
       "16. 'V16'\n",
       "17. 'V17'\n",
       "18. 'V18'\n",
       "19. 'V19'\n",
       "20. 'V20'\n",
       "21. 'V21'\n",
       "22. 'V22'\n",
       "23. 'V23'\n",
       "24. 'V24'\n",
       "25. 'V25'\n",
       "26. 'V26'\n",
       "27. 'V27'\n",
       "28. 'V28'\n",
       "29. 'V29'\n",
       "30. 'V30'\n",
       "31. 'V31'\n",
       "32. 'V32'\n",
       "33. 'V33'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"all.ids\" \"V2\"      \"V3\"      \"V4\"      \"V5\"      \"V6\"      \"V7\"     \n",
       " [8] \"V8\"      \"V9\"      \"V10\"     \"V11\"     \"V12\"     \"V13\"     \"V14\"    \n",
       "[15] \"V15\"     \"V16\"     \"V17\"     \"V18\"     \"V19\"     \"V20\"     \"V21\"    \n",
       "[22] \"V22\"     \"V23\"     \"V24\"     \"V25\"     \"V26\"     \"V27\"     \"V28\"    \n",
       "[29] \"V29\"     \"V30\"     \"V31\"     \"V32\"     \"V33\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output.dt <- data.table(output)\n",
    "names(output.dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'person_id'</li><li>'PC1'</li><li>'PC2'</li><li>'PC3'</li><li>'PC4'</li><li>'PC5'</li><li>'PC6'</li><li>'PC7'</li><li>'PC8'</li><li>'PC9'</li><li>'PC10'</li><li>'PC11'</li><li>'PC12'</li><li>'PC13'</li><li>'PC14'</li><li>'PC15'</li><li>'PC16'</li><li>'PC17'</li><li>'PC18'</li><li>'PC19'</li><li>'PC20'</li><li>'PC21'</li><li>'PC22'</li><li>'PC23'</li><li>'PC24'</li><li>'PC25'</li><li>'PC26'</li><li>'PC27'</li><li>'PC28'</li><li>'PC29'</li><li>'PC30'</li><li>'PC31'</li><li>'PC32'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'person\\_id'\n",
       "\\item 'PC1'\n",
       "\\item 'PC2'\n",
       "\\item 'PC3'\n",
       "\\item 'PC4'\n",
       "\\item 'PC5'\n",
       "\\item 'PC6'\n",
       "\\item 'PC7'\n",
       "\\item 'PC8'\n",
       "\\item 'PC9'\n",
       "\\item 'PC10'\n",
       "\\item 'PC11'\n",
       "\\item 'PC12'\n",
       "\\item 'PC13'\n",
       "\\item 'PC14'\n",
       "\\item 'PC15'\n",
       "\\item 'PC16'\n",
       "\\item 'PC17'\n",
       "\\item 'PC18'\n",
       "\\item 'PC19'\n",
       "\\item 'PC20'\n",
       "\\item 'PC21'\n",
       "\\item 'PC22'\n",
       "\\item 'PC23'\n",
       "\\item 'PC24'\n",
       "\\item 'PC25'\n",
       "\\item 'PC26'\n",
       "\\item 'PC27'\n",
       "\\item 'PC28'\n",
       "\\item 'PC29'\n",
       "\\item 'PC30'\n",
       "\\item 'PC31'\n",
       "\\item 'PC32'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'person_id'\n",
       "2. 'PC1'\n",
       "3. 'PC2'\n",
       "4. 'PC3'\n",
       "5. 'PC4'\n",
       "6. 'PC5'\n",
       "7. 'PC6'\n",
       "8. 'PC7'\n",
       "9. 'PC8'\n",
       "10. 'PC9'\n",
       "11. 'PC10'\n",
       "12. 'PC11'\n",
       "13. 'PC12'\n",
       "14. 'PC13'\n",
       "15. 'PC14'\n",
       "16. 'PC15'\n",
       "17. 'PC16'\n",
       "18. 'PC17'\n",
       "19. 'PC18'\n",
       "20. 'PC19'\n",
       "21. 'PC20'\n",
       "22. 'PC21'\n",
       "23. 'PC22'\n",
       "24. 'PC23'\n",
       "25. 'PC24'\n",
       "26. 'PC25'\n",
       "27. 'PC26'\n",
       "28. 'PC27'\n",
       "29. 'PC28'\n",
       "30. 'PC29'\n",
       "31. 'PC30'\n",
       "32. 'PC31'\n",
       "33. 'PC32'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"person_id\" \"PC1\"       \"PC2\"       \"PC3\"       \"PC4\"       \"PC5\"      \n",
       " [7] \"PC6\"       \"PC7\"       \"PC8\"       \"PC9\"       \"PC10\"      \"PC11\"     \n",
       "[13] \"PC12\"      \"PC13\"      \"PC14\"      \"PC15\"      \"PC16\"      \"PC17\"     \n",
       "[19] \"PC18\"      \"PC19\"      \"PC20\"      \"PC21\"      \"PC22\"      \"PC23\"     \n",
       "[25] \"PC24\"      \"PC25\"      \"PC26\"      \"PC27\"      \"PC28\"      \"PC29\"     \n",
       "[31] \"PC30\"      \"PC31\"      \"PC32\"     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setnames(output.dt, c(\"person_id\", paste0(\"PC\",seq(1:32))))\n",
    "names(output.dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head(output.dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile <- \"myPCA-2023-02-13.csv\"\n",
    "write.csv(output.dt, outfile,quote=FALSE, row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'gsutil -m cp -r myPCA-2023-02-13.csv gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/'"
      ],
      "text/latex": [
       "'gsutil -m cp -r myPCA-2023-02-13.csv gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/'"
      ],
      "text/markdown": [
       "'gsutil -m cp -r myPCA-2023-02-13.csv gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/'"
      ],
      "text/plain": [
       "[1] \"gsutil -m cp -r myPCA-2023-02-13.csv gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#(cp.command <- paste0(\"gsutil -m cp -r \", outfile,\" gs://fc-secure-30fdbdfd-a46b-406d-9617-1bc69ae1da9d/data/PCSNPS/\"))\n",
    "(cp.command <- paste0(\"gsutil -m cp -r \", outfile,\" gs://fc-secure-91b7b957-58d8-4142-b93b-c3a2635f2d78/\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(cp.command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "snpgdsClose(genofile)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
